File: Day323_2025/Q323_GFG.java
import java.util.Arrays;
import java.util.PriorityQueue;

class Q323_LeetCode {
    public int minCostPath(int[][] heights) {
        int n = heights.length;
        int m = heights[0].length;
        
        int[][] dist = new int[n][m];
        for (int i = 0; i < n; i++) 
            Arrays.fill(dist[i], (int)1e9);

        PriorityQueue<int[]> pq = new PriorityQueue<>(
            (a, b) -> Integer.compare(a[0], b[0])  // {effort, r, c}
        );

        pq.add(new int[]{0, 0, 0});
        dist[0][0] = 0;

        int[] dx = {1, -1, 0, 0};
        int[] dy = {0, 0, 1, -1};

        while (!pq.isEmpty()) {
            int[] curr = pq.poll();
            int effort = curr[0];
            int r = curr[1];
            int c = curr[2];

            if (r == n - 1 && c == m - 1) return effort;

            for (int i = 0; i < 4; i++) {
                int nr = r + dx[i];
                int nc = c + dy[i];

                if (nr >= 0 && nr < n && nc >= 0 && nc < m) {
                    int newEffort = Math.max(
                        Math.abs(heights[r][c] - heights[nr][nc]),
                        effort
                    );

                    if (dist[nr][nc] > newEffort) {
                        dist[nr][nc] = newEffort;
                        pq.add(new int[]{newEffort, nr, nc});
                    }
                }
            }
        }
        return 0;
    }
}

File: Day323_2025/Q323_GFG.py
import heapq

class Solution:
    def minCostPath(self, heights):
        n = len(heights)
        m = len(heights[0])

        dist = [[10**9] * m for _ in range(n)]
        dist[0][0] = 0

        pq = [(0, 0, 0)]  # (effort, r, c)

        dx = [1, -1, 0, 0]
        dy = [0, 0, 1, -1]

        while pq:
            effort, r, c = heapq.heappop(pq)

            if r == n - 1 and c == m - 1:
                return effort

            for i in range(4):
                nr = r + dx[i]
                nc = c + dy[i]

                if 0 <= nr < n and 0 <= nc < m:
                    new_effort = max(
                        abs(heights[r][c] - heights[nr][nc]),
                        effort
                    )

                    if dist[nr][nc] > new_effort:
                        dist[nr][nc] = new_effort
                        heapq.heappush(pq, (new_effort, nr, nc))
